monad-rpc/src/chainstate/mod.rs

```diff
@@ -781,6 +813,83 @@ impl<T: Triedb> ChainState<T> {
     }
 }
 
+async fn try_collect_logs_stream_with_heuristic_response_limit<E>(
+    max_response_size: u32,
+    from_block: u64,
+    to_block: u64,
+    stream: impl Stream<Item = Result<(u64, impl IntoIterator<Item = Log>), E>>,
+) -> JsonRpcResult<Result<Vec<MonadLog>, E>> {
+    let mut stream = std::pin::pin!(stream);
+
+    const EXTRAPOLATION_CHECK_MIN_RESPONSE_SIZE: u64 = 4 * 1024 * 1024;
+    const EXTRAPOLATION_CHECK_MIN_BLOCKS: u64 = 100;
+
+    let num_blocks_total = to_block + 1 - from_block;
+
+    let mut num_blocks_processed = 0u64;
+    let mut heuristic_response_size = 0u64;
+
+    let mut response_logs = Vec::<MonadLog>::default();
+
+    while let Some(result) = stream.next().await {
+        match result {
+            Err(err) => return Ok(Err(err)),
+            Ok((block_number, logs)) => {
+                if block_number != from_block.saturating_add(num_blocks_processed) {
+                    error!(
+                        ?from_block,
+                        ?num_blocks_processed,
+                        ?block_number,
+                        "logs stream block numbers inconsistent"
+                    );
+                    return Err(JsonRpcError::internal_error(format!("Logs out of order")));
+                }
+
+                num_blocks_processed += 1;
+
+                if num_blocks_processed > num_blocks_total {
+                    error!(
+                        ?from_block,
+                        ?num_blocks_processed,
+                        ?block_number,
+                        ?num_blocks_total,
+                        "logs stream block number exceeded range"
+                    );
+                    return Err(JsonRpcError::internal_error(format!("Logs out of range")));
+                }
+
+                response_logs.extend(logs.into_iter().map(|log| {
+                    heuristic_response_size += compute_heuristic_log_len(&log) as u64;
+                    MonadLog(log)
+                }));
+
+                if heuristic_response_size > max_response_size as u64 {
+                    return Err(JsonRpcError::max_size_exceeded());
+                }
+
+                if heuristic_response_size >= EXTRAPOLATION_CHECK_MIN_RESPONSE_SIZE
```

Comment:
good example of where a comment explaining the rule and a justification for why this rule is reasonable would be good documentation